src/gui/guiFormSpecMenu.cpp

```diff
@@ -4228,153 +4290,370 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
+				else if (button == BET_MIDDLE)
+					count = MYMIN(s_count, 10);
+				else if (button == BET_WHEEL_DOWN)
+					count = 1;
+				else if (button == BET_LEFT)
+					count = s_count;
+
+				if (event.MouseInput.Shift) {
+					// Shift pressed: move item, right click moves 1
+					shift_move_amount = button == BET_RIGHT ? 1 : count;
+				} else {
+					// No shift: select item
 					m_selected_item = new GUIInventoryList::ItemSpec(s);
-
-					u32 count;
-					if (button == BET_RIGHT)
-						count = (s_count + 1) / 2;
-					else if (button == BET_MIDDLE)
-						count = MYMIN(s_count, 10);
-					else if (button == BET_WHEEL_DOWN)
-						count = 1;
-					else  // left
-						count = s_count;
-
-					if (!event.MouseInput.Shift) {
-						// no shift: select item
-						m_selected_amount = count;
-						m_selected_dragging = button != BET_WHEEL_DOWN;
-						m_auto_place = false;
-					} else {
-						// shift pressed: move item, right click moves 1
-						shift_move_amount = button == BET_RIGHT ? 1 : count;
-					}
+					m_selected_amount = count;
+					m_selected_dragging = button != BET_WHEEL_DOWN;
 				}
-			} else { // m_selected_item != NULL
-				assert(m_selected_amount >= 1);
 
-				if (s.isValid()) {
-					// Clicked a slot: move
-					if (button == BET_RIGHT || button == BET_WHEEL_UP)
-						move_amount = 1;
-					else if (button == BET_MIDDLE)
-						move_amount = MYMIN(m_selected_amount, 10);
-					else if (button == BET_LEFT)
-						move_amount = m_selected_amount;
-					// else wheeldown
+			} else if (m_selected_item) {
+				// Clicked a slot: move
+				if (button == BET_RIGHT || button == BET_WHEEL_UP)
+					move_amount = 1;
+				else if (button == BET_WHEEL_DOWN)
+					pickup_amount = MYMIN(s_count, 1);
+				else if (button == BET_MIDDLE)
+					move_amount = MYMIN(m_selected_amount, 10);
+				else if (button == BET_LEFT)
+					move_amount = m_selected_amount;
+
+				if (event.MouseInput.Shift && !identical && matching) {
+					// Shift-move all items the same as the selected item to the next list
+					move_amount = 0;
+
+					// Try to find somewhere to move the items to
+					s16 r = getNextInventoryRing(s.inventoryloc, s.listname);
+					if (r < 0) // Not found
+						break;
 
-					if (identical) {
-						if (button == BET_WHEEL_DOWN) {
-							if (m_selected_amount < s_count)
-								++m_selected_amount;
-						} else {
-							if (move_amount >= m_selected_amount)
-								m_selected_amount = 0;
-							else
-								m_selected_amount -= move_amount;
-							move_amount = 0;
+					const ListRingSpec &to_ring = m_inventory_rings[r];
+					Inventory *inv_to = m_invmgr->getInventory(to_ring.inventoryloc);
+					if (!inv_to)
+						break;
+					InventoryList *list_to = inv_to->getList(to_ring.listname);
+					if (!list_to)
+						break;
+
+					ItemStack slct = list_selected->getItem(m_selected_item->i);
+
+					for (u32 i = 0; i < list_s->getSize(); i++) {
+						// Skip the selected slot
+						if (i == m_selected_item->i)
```

Comment:
`src/gui/guiFormSpecMenu.cpp:4382:55: warning: comparison of integer expressions of different signedness: ‘irr::u32’ {aka ‘unsigned int’} and ‘irr::s32’ {aka ‘int’} [-Wsign-compare]`

I suppose you could use `s32` to iterate unless you'd like to use `u32` in `GUIInventoryList::ItemSpec`.